### PR TITLE
Fix iterbatches not iterating in batch_size across shards and pad_batches with None y or w.

### PR DESCRIPTION
**Pull request by proteneer**
This PR:

[x] Addresses #911 
[x] Uses carries to so that we can iterate in consistent batch_sizes even across shards
[x] Adds documentation to iterabatches
[x] Adds significantly more stringent tests to iterbatches to guarantee the invariant
[x] Allows shuffling of the carry as well
[x] Add an extra test for pad_batches in iterbatches
[x] Preserves behavior of batch_size=None to continue iterating in shard_size
[x] Add an extra test for iterating over None y or w and pad_batches is True
[x] Fix pad_batches when y or w is None
[x] yapf

Notably, we now guarantee that the number of batches iterated is strictly equal to math.ceil(dataset.get_shape()[0][0])/batch_size.

### CODE REVIEW

**Comment by proteneer** [comment]
@rbharath @peastman  @lilleswing Can you guys help me sanity check this? This is a super important method that's used pretty much everywhere so I want to be sure I'm not doing something stupid.

File: deepchem/data/datasets.py [hunk 1]
```diff
@@ -636,9 +637,30 @@ def iterbatches(self,
                   epoch=0,
                   deterministic=False,
                   pad_batches=False):
-    """Get an object that iterates over minibatches from the dataset.
-
+    """ Get an object that iterates over minibatches from the dataset. It is guaranteed
+    that the number of batches returned is math.ceil(dataset.get_shape()[0][0]/batch_size).
```
Comment (rbharath) [inline deepchem/data/datasets.py]: Maybe just say `len(dataset)/batch_size`. I believe this means the same thing and is more readable.

File: deepchem/data/tests/test_datasets.py [hunk 2]
```diff
@@ -395,6 +397,109 @@ def test_disk_iterate_batch_size(self):
       batch_sizes.append(len(X))
     self.assertEqual([3, 3, 3, 1], batch_sizes)
 
+  def test_disk_iterate_batch(self):
```
Comment (rbharath) [inline deepchem/data/tests/test_datasets.py]: Could you break this into multiple simpler tests? It's hard to follow along with the test given the length.

File: deepchem/data/tests/test_datasets.py [hunk 3]
```diff
@@ -395,6 +397,109 @@ def test_disk_iterate_batch_size(self):
...
+
+    all_batch_sizes = [
```
Comment (peastman) [inline deepchem/data/tests/test_datasets.py]: `all_batch_sizes` gets created, but then it doesn't look like it ever gets used.

**Comment by peastman** [comment]
I don't see any problems with the implementation, but the logic is complicated enough that I'm not confident I haven't missed something.  We just need to test it really thoroughly.

I think there may be a bug in the test case.  You generate a list of different batch sizes to try, but then ignore that list and just do everything with whatever was the last size in the list.

File: deepchem/data/tests/test_datasets.py (hunk 2 above)
Comment (peastman) [reply deepchem/data/tests/test_datasets.py]: All it really does is create a dataset, then iterate over it in both deterministic and nondeterministic order and make sure each one gives the right results.  It repeats that a bunch of times for random patterns of shard sizes.  It could maybe be simplified a bit, but I don't think it makes sense to break it into multiple tests.

File: deepchem/data/datasets.py (hunk 1 above)
Comment (proteneer) [reply deepchem/data/datasets.py]: Thanks - will fix this

**Comment by proteneer** [comment]
@peastman Thanks - I fixed the bug in the test case.

File: deepchem/data/datasets.py [hunk 4]
```diff
@@ -636,65 +637,127 @@ def iterbatches(self,
...
+      pool = Pool(1)  # mp.dummy aliases ThreadPool to Pool
       next_shard = pool.apply_async(dataset.get_shard, (shard_perm[0],))
-      for i in range(num_shards):
+
+      total_yield = 0
+
+      if batch_size is None:
+        num_global_batches = num_shards
+      else:
+        num_global_batches = math.ceil(len(dataset) / batch_size)
```
Comment (lilleswing) [inline deepchem/data/datasets.py]: Note to self
```python
len(dataset) # literally reads in all the data
```

Comment (lilleswing) [reply deepchem/data/datasets.py]: We should implement some caching -- not part of this review.  I can store these things in the side car json file that I am making as part of the transition to a stable store.

File: deepchem/data/datasets.py [hunk 5]
```diff
@@ -636,65 +637,127 @@ def iterbatches(self,
...
+          next_shard = pool.apply_async(dataset.get_shard,
+                                        (shard_perm[cur_shard + 1],))
         else:
-          sample_perm = np.arange(n_samples)
+          pool.close()
+
+        if carry is not None:
+          X = np.concatenate([carry[0], X], axis=0)
+          y = np.concatenate([carry[1], y], axis=0)
+          w = np.concatenate([carry[2], w], axis=0)
```
Comment (lilleswing) [inline deepchem/data/datasets.py]: do we have to do None checks here?
there are cases when y and w are None.

File: deepchem/data/datasets.py (hunk 4 above)
Comment (proteneer) [reply deepchem/data/datasets.py]: wait what? so I should revert to get_shape()[0][0] - afaik that just looks up the metadata

File: deepchem/data/datasets.py (hunk 5 above)
Comment (proteneer) [reply deepchem/data/datasets.py]: @lilleswing you're right about this. Will add a guard.

Comment (proteneer) [reply deepchem/data/datasets.py]: Oddly enough even pad_batches doesn't guard against this:

``` python
def pad_batch(batch_size, X_b, y_b, w_b, ids_b):
  """Pads batch to have size precisely batch_size elements.

  Fills in batch by wrapping around samples till whole batch is filled.
  """
  num_samples = len(X_b)
  if num_samples == batch_size:
    return (X_b, y_b, w_b, ids_b)
  # By invariant of when this is called, can assume num_samples > 0
  # and num_samples < batch_size
  if len(X_b.shape) > 1:
    feature_shape = X_b.shape[1:]
    X_out = np.zeros((batch_size,) + feature_shape, dtype=X_b.dtype)
  else:
    X_out = np.zeros((batch_size,), dtype=X_b.dtype)

  num_tasks = y_b.shape[1] # fucked if None
  y_out = np.zeros((batch_size, num_tasks), dtype=y_b.dtype)
  w_out = np.zeros((batch_size, num_tasks), dtype=w_b.dtype) # fucked if None
  ids_out = np.zeros((batch_size,), dtype=ids_b.dtype)

  # Fill in batch arrays
  start = 0
  # Only the first set of copy will be counted in training loss
  w_out[start:start + num_samples] = w_b[:]
  while start < batch_size:
    num_left = batch_size - start
    if num_left < num_samples:
      increment = num_left
    else:
      increment = num_samples
    X_out[start:start + increment] = X_b[:increment]
    y_out[start:start + increment] = y_b[:increment]
    ids_out[start:start + increment] = ids_b[:increment]
    start += increment

  return (X_out, y_out, w_out, ids_out)
```

Comment (proteneer) [reply deepchem/data/datasets.py]: @lilleswing Is there ever a use case where pad_batch=True and y/w could be None?

Comment (peastman) [reply deepchem/data/datasets.py]: I don't think they're related.  If the dataset doesn't include weights, that's a property of the dataset.  `pad_batches` is an option determined by the calling code.

**Comment by proteneer** [comment]
Tests passed. I think this is ready to merge. @rbharath

**Comment by rbharath** [comment]
LGTM
